modules/adwords_appengine/src/main/java/com/google/api/ads/adwords/jaxws/utils/v201802/ServiceQuery.java

```diff
@@ -0,0 +1,290 @@
+import com.google.api.ads.adwords.jaxws.v201802.cm.AdGroupBidLandscapePage;
+import com.google.api.ads.adwords.jaxws.v201802.cm.CriterionBidLandscape;
+import com.google.api.ads.adwords.jaxws.v201802.cm.CriterionBidLandscapePage;
+import com.google.api.ads.adwords.jaxws.v201802.cm.Page;
+import com.google.api.ads.adwords.jaxws.v201802.cm.SortOrder;
+import com.google.api.ads.adwords.lib.selectorfields.EntityField;
+import com.google.api.ads.adwords.lib.utils.AdWordsInternals;
+import com.google.api.ads.adwords.lib.utils.ServiceQueryInterface;
+import com.google.api.ads.common.lib.utils.AdsUtilityInvocationHandler;
+import com.google.common.base.Objects;
+import com.google.common.reflect.Reflection;
+import java.lang.reflect.InvocationHandler;
+import javax.annotation.Nullable;
+
+/** A service query that embodies AWQL string for making SOAP requests. */
+public final class ServiceQuery implements ServiceQueryInterface<Page> {
+
+  private final String partialAwqlQuery;
+  private final int pageSize;
+  private int startIndex;
+
+  /**
+   * Constructs a service query object using the passed partial AWQL query, the start index and the
+   * page size.
+   *
+   * <p>This constructor should be called by {@link Builder} only.
+   *
+   * @param partialAwqlQuery the partial AWQL query without the LIMIT clause, created by the service
+   *     query builder
+   * @param startIndex the start index for the LIMIT clause
+   * @param pageSize the page size for the LIMIT clause
+   */
+  ServiceQuery(String partialAwqlQuery, int startIndex, int pageSize) {
+    this.partialAwqlQuery = partialAwqlQuery;
+    this.startIndex = startIndex;
+    this.pageSize = pageSize;
+  }
+
+  @Override
+  public ServiceQuery nextPage(@Nullable Page page) {
+    return (page == null) ? this : nextPage(this.pageSize);
+  }
+
+  /**
+   * Sets the LIMIT clause of the AWQL to the next page calculated by using the total number of
+   * landscape points in {@link AdGroupBidLandscapePage}. If {@code page} is null, this will not
+   * change the LIMIT clause.
+   *
+   * <p>This method is meant to be used with {@code hasNext()}. The {@code page} is necessary when
+   * using DataService, as its paging mechanism is different from other services. For details, see
+   * https://developers.google.com/adwords/api/docs/guides/bid-landscapes#paging_through_results.
+   *
+   * @param page the ad group bid landscape page whose total number of landscape points will be
+   *     used to increment the index in the LIMIT clause
+   * @return this service query object
+   */
+  public ServiceQuery nextPage(@Nullable AdGroupBidLandscapePage page) {
+    if (page == null) {
+      return this;
+    }
+    int totalLandscapePointsInPage = 0;
+    for (AdGroupBidLandscape adGroupBidLandscape : page.getEntries()) {
+      totalLandscapePointsInPage += adGroupBidLandscape.getLandscapePoints().size();
+    }
+    return nextPage(totalLandscapePointsInPage);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 77:84, 101:108.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.